Rewrite throttle logic to prevent request flooding

kkoomen opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 1303


## What is the new behavior?

I've made some adjustments based on my suggestion in #1303 in order to add a single record and retrieve all info in a single function. This also allows external storage providers made by the community to add and retrieve in a single roundtrip (i.e. my [redis storage provider](https://github.com/kkoomen/nestjs-throttler-storage-redis)).

The initial behavior was that `getRecord()` returned an list of sorted TTL timestamps, then if it didn't reach the limit, it will call `addRecord()`. This change was made based on Redis where I fiddled around on how to prevent this issue. I found out that [express-rate-limit](https://github.com/express-rate-limit/express-rate-limit) is incrementing a single number and returning the information in a single roundtrip, which is significantly faster than how NestJS throttler works by called `getRecord()`, then `addRecord`.

The [rate-limit-redis](https://github.com/wyattjoh/rate-limit-redis) did have a very fast version with `EVALSHA`. I tried to integrate the `EVALSHA` script they used into my own storage provider, which immediately worked. Then I adjust the changes to my version, but that did not work all the way. There were still coming +10 requests through with my version. This version was much faster than before where ±400 requests came through, but still wasn't not perfect. This was because `SMEMBERS <key>` that I personally used - which returns a list of TTL timestamps from `<key>` - seems to be much slower than it is to increment a single number and retrieve this number. Therefore, I made the new logic similar to how [express-rate-limit](https://github.com/express-rate-limit/express-rate-limit) works by requiring two things from the `addRecord()` (which I have renamed in this PR): 
- `totalHits`: amount of requests done for `key`
- `timeToExpire`: amount of seconds until the `totalHits` expire for `key`

## Does this PR introduce a breaking change?
- [x] Yes
- [ ] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

Breaking `ThrottlerStorage` changes:
- removed `getRecord`
- `addRecord(key: string, ttl: number): Promise<number[]>;` changes to `increment(key: string, ttl: number): Promise<ThrottlerStorageRecord>;`

## Other information

kkoomen commented:
@jmcdo29 I still feel like we should put `{ totalHits: number; timeToExpire: number }` in a variable, what do you think?

jmcdo29 commented on README.md:
```diff
@@ -216,8 +215,8 @@ The interface looks like this:
 export interface ThrottlerStorage {
-  getRecord(key: string): Promise<number[]>;
-  addRecord(key: string, ttl: number): Promise<void>;
+  storage: Record<string, { totalHits: number; timeToExpire: number }>;
+  addRecord(key: string, ttl: number): Promise<{ totalHits: number; timeToExpire: number }>;
```
Yeah, we can make an interface for `{ totalHits, timeToExpire }`. Also make sure to export it for the sake of extension packages

kkoomen replied:
I've updated this part as well as other part of the README. Have a look again and resolve this is you're satisfied.

jmcdo29 reviewed (commented):
Overall code looks good and clean. Just add that interface.

kkoomen commented:
@jmcdo29 I have adjusted some small issues and have refactored some of the code. The things to notice is:
- `timeToExpire` will be reset in the ThrottlerStorageService once it is 0 or less (this was already done in my redis storage provider, but I forgot to add it here)
- two new interfaces for the sake of extensibility: `ThrottlerStorageOptions` and `ThrottlerStorageRecord` (the latter will be very useful for my own package as well, the former is more for internal use of this package, but is still nice to have because we obviously don't want to repeat ourselves as developers :))
- renamed `addRecord` to `increment`: I felt the `addRecord` was not the main purpose of the function as we're incrementing the `totalHits` as of now.
- Adjusted the regex in `test/controller.e2e-spec.ts` for `x-ratelimit-reset` so that we only require whole numbers. The previous version was only checking if it contained 1 or more numbers, which isn't what we should check, because that means any arbitrary string can come through, as long as it has a single digit in it.

jmcdo29 commented on src/throttler-storage.interface.ts:
```diff
@@ -3,14 +3,17 @@ export interface ThrottlerStorage {
+   *   - expiresAt: Epoch milliseconds when this key expires.
    */
-  storage: Record<string, { totalHits: number; timeToExpire: number }>;
+  storage: Record<string, { totalHits: number; expiresAt: number }>;
```
With how often we use the `{ totalHits: number, expiresAt: number }` type, should we make an interface for it? `ThrottlerData` or something like that?

jmcdo29 replied:
Ignore this. GitHub brought me to an old version of the PR

jmcdo29 commented:
Everything looks good to me here. 

@kamilmysliwiec this would be a part of a "breaking" change in that standard use of the library should not change at all, but the underlying API that developers could extend and work with for custom storages would change. I wanted to bring this to your attention before publishing a major release for it in case you noticed anything you wanted addressed.

nadavkaner commented:
Hey do you have an ETA on merging this PR?

nadavkaner commented:
@jmcdo29 Hey do you have an ETA on merging this PR?

jmcdo29 commented:
> @jmcdo29 Hey do you have an ETA on merging this PR?

I'll likely get this merged and published later this week

nadavkaner commented:
Thanks that will be great, we noticed some issues on requests spike which related to this issue

zhorakaroy commented:
Hi @jmcdo29 , any news about merging PR?

jmcdo29 commented:
I'll get the docs change started up tomorrow and get the ball rolling on all of that

jmcdo29 commented:
Welp, that rebase mucked things up :sweat_smile: I'll get that resolved and merged today.

jmcdo29 commented:
@kkoomen heads up, I had to force push to your branch to remove most of those renovate/dependabot commits. All should be clean now though. Just a heads up for next time you work on this library unless you just delete and re-fork.

kkoomen commented:
@jmcdo29 Thanks, I'll start merging things on my side.
